core/string/ustring.cpp

```diff
@@ -1447,15 +1447,27 @@ String String::num(double p_num, int p_decimals) {
 		fmt[5] = 'f';
 		fmt[6] = 0;
 	}
-	char buf[256];
+	// if we want to convert a double with as much decimal places as as DBL_MAX
+	// or DBL_MIN then we would theoretically need a buffer of at least
+	// DBL_MAX_10_EXP + 2 for DBL_MAX and DBL_MAX_10_EXP + 4 for DBL_MIN.
+	// BUT those values where still giving me exceptions, so I tested from
+	// DBL_MAX_10_EXP + 10 incrementing one by one and DBL_MAX_10_EXP + 17 was
+	// the first buffer size not to throw an exception
+	char buf[DBL_MAX_10_EXP + 17];
+
+
 
 #if defined(__GNUC__) || defined(_MSC_VER)
-	snprintf(buf, 256, fmt, p_num);
+	// PLEASE NOTE that, albeit vcrt online reference states that snprintf
+	// should safely truncate the output to the given buffer size, we have
+	// found a case where this is not true, so we should create a buffer
+	// as big as needed
+	snprintf(buf, DBL_MAX_10_EXP + 17, fmt, p_num);
```

Comment:
I wasn't able to test your snippet
![image](https://user-images.githubusercontent.com/5783414/195090933-0904642a-6bff-4e43-bbae-c6dbd6898f86.png)
_snprintf_s expects as char* const as first argument
![image](https://user-images.githubusercontent.com/5783414/195091294-94e12951-1e4a-424f-b47b-ba99c9acfa9b.png)
but passing nullptr, or NULL, with or without explicit casting to char* or char* const yields an error.

Reply:
That's strange, `_snprintf_s` have a different number of arguments than `snprintf` in can't be used as a replacement. What kind of the compiler are you using?

On the other hand, snippet is a bit wrong, it should be something like:

```c++
int sz = snprintf(nullptr, 0, fmt, p_num);
char *buf = static_cast<char *>(memalloc(sz + 1));
snprintf(buf, sz + 1, fmt, p_num);
...
memfree(buf);
```

Reply:
there are two versions of snprintf
![Screenshot (192)](https://user-images.githubusercontent.com/5783414/195098506-44805556-55ce-4942-8143-bbaba196fb9c.png)
![Screenshot (193)](https://user-images.githubusercontent.com/5783414/195098508-1c2814cb-ce61-4d49-ba43-ea559556eb8f.png)

Reply:
I wasnt able to use your snippet with the first version, only the second. and even so I had to use alloca (or could alocate on the heap like you did with malloc) but it didnt work with a null pointer for the buffer

Reply:
I just tried it with MSVC 2022, and it worked fine. Is it not compiling, or just an IDE showing an error underscore (it might detect Godot config defines incorrectly)?

Reply:
I fails compiling both on the IDE and on the command line. Also that second snprintf call cannot be made with a char*, only with a char&. The version that accepts char* is the one that gets both the buffer size and maximum character count.
![Screenshot (195)](https://user-images.githubusercontent.com/5783414/195102493-2cc54789-e908-476f-ab0b-bcdc29343b6d.png)

Reply:
both _snprintf_s implementations come from the same place, stdio.h in c:\program files\windows kits\10\include\10.0.19041.0\ucrt\stdio.h
![Screenshot (197)](https://user-images.githubusercontent.com/5783414/195104949-cb7498d5-cbd7-409e-905a-b9b4184b6ced.png)

Reply:
There's no version of `snprintf` that have max character count - https://en.cppreference.com/w/c/io/fprintf, it's `_snprinf_s` that takes it. Not sure what's going on, but it might be something wrong with your Windows SDK headers setup.

And I also can't reproduce the issue, MSVC 2022 `snprintf` seems to be truncation string correctly. The issues with the SDK might also explain the crash, if it's somehow using `_snprinf_s` with a different number of arguments.

Reply:
For the reference, I'm using VS 2022 - 17.3.5 on Windows 11 (22H2).

Reply:
I'm using visual studio 17.0.1 on windows 10

Let's assume it is indeed a problem with my personal setup. That put aside, let me then advocate in favor of increasing the buffer size.

The String::num function is meant to convert a double to a string representation. It allows specifying the number of decimal places, but it does not impose (explicitly or implicitly) a limit for the integer places of a real number. This double may be any double, even as large as DBL_MAX. If we let the buffer size be 256 as it currently is, even though very large doubles won't crash since we assume vcrt snprintf will truncate safely, the return value string will not represent the input value double. The user will expect a string with about 308 characters, but will get only 255. And since it is in fact an integer (since it is larger than 2^53), the returned value (as string) will represent a very different value from the input, wich will be several orders of magnitude greater.

Reply:
Increasing buffer size definitely is a correct thing to do. The only thing that I do not like is `+ 17` which is a strange value with unknown meaning which might be confusing in the future.

Reply:
As requested, I have replaced the confusing "DBL_MAX_10_EXP + 17" expression with a round value.